### DISCUSSION TITLE
Downloading a PDF file from an Endpoint

### DISCUSSION THREAD

**Discussion by nick-goloborodko** [jasontaylordev/CleanArchitecture#1101] [General]
This is driving me mad. Seems like such a simple task - but I just can't seem to figure out how to accomplish with .NET 8 CA solution.

I have a PDF file that sits on a protected cloud storage. I need to develop a method that takes that and returns the file contents with mime type . The issues with the code below is that it returns data as JSON, with file contents base64 encoded, rather than the http response itself being of the correct mime type and just the file contents in the body of the response?

```csharp
    public async Task<IActionResult> DownloadTemplate(ISender sender, long id)
    {
        var data = await sender.Send(new DownloadTemplateQuery { Id = id });
        return new FileContentResult(data.Data, data.MimeType);

    }
```

I used to be able to do something like this in MVC controllers, but this approach no longer appears to work?
```csharp
return File(document.Data, document.ContentType, document.Name);
```

Any help will be greatly appreciated!

**Comment by Nizar-Almshantaf** [comment]
any update on this ? did you manage to solve it ? when using "new FileContentResult(data.Data, data.MimeType);" I'm getting a json response with fileContents encoded in base64, which previously I used to get a download button from swagger and used to receive the file as binary

**Comment by samabos** [comment]
You can try:
sample.pdf file should be in root directory. set the file property to `Copy Always` to test
```csharp

    public async Task<IResult> ReadAndReturnFile()
    {
        // Get the full path to the file in the root directory
        var fullPath = Path.Combine(AppContext.BaseDirectory, "sample.pdf");

        if (!File.Exists(fullPath))
        {
            return Results.NotFound("File not found.");
        }

        // Read file contents
        var fileBytes = File.ReadAllBytes(fullPath);

        // Return the file as a response
        return Results.File(fileBytes, "application/pdf", Path.GetFileName(fullPath));
    }
```
